### PR TITLE
World Map clustering

### PR DESCRIPTION
**Pull request by lawik**
See video here: https://github.com/nerves-hub/nerves_hub_web/pull/1611

I tested this with 180.000 active and connected devices (faked). It loaded consistently but took a decent number of seconds. It was much faster down towards 50.000 and snappy below those types of numbers.

Once loaded it becomes easy to navigate and performant. I think this works. We could look at faster options. But the clustering makes 180K possible where it isn't without it.

We could try various things to constrain how many we load. I've looked at recommendations for performance and applied what I found. chunkedLoading option keeps it responsive while loading, renderer using Canvas makes it more performant.

I cut down heavily on non-used values because the first JSON transfer was 18 Mb and I think I got it down to about 1Mb for 180K. Maybe we could win more there with compression. That wasn't the slow part for me.

### CODE REVIEW

File: mix.lock [hunk 1]
```diff
@@ -52,23 +51,23 @@
...
-  "phoenix_test": {:hex, :phoenix_test, "0.3.1", "adf5d67cb152fa1e0220527a9827db7f865a20817c7c0161315ba6fe86a8946c", [:mix], [{:deep_merge, "~> 1.0", [hex: :deep_merge, repo: "hexpm", optional: false]}, {:floki, ">= 0.30.0", [hex: :floki, repo: "hexpm", optional: false]}, {:jason, "~> 1.4", [hex: :jason, repo: "hexpm", optional: false]}, {:phoenix, "~> 1.7.10", [hex: :phoenix, repo: "hexpm", optional: false]}, {:phoenix_live_view, "~> 0.20.1", [hex: :phoenix_live_view, repo: "hexpm", optional: false]}], "hexpm", "03f24332a966673ff40d35c45f427c7671537ea508b5777141dd60e60cdae22a"},
+  "phoenix_test": {:hex, :phoenix_test, "0.3.2", "030d98c401ef95db564da6a15b5077909e11a532986690b757f40c105cda8be4", [:mix], [{:floki, ">= 0.30.0", [hex: :floki, repo: "hexpm", optional: false]}, {:jason, "~> 1.4", [hex: :jason, repo: "hexpm", optional: false]}, {:phoenix, "~> 1.7.10", [hex: :phoenix, repo: "hexpm", optional: false]}, {:phoenix_live_view, "~> 0.20.1", [hex: :phoenix_live_view, repo: "hexpm", optional: false]}], "hexpm", "538be4b25207e250b1939f6374bb1fc9f6346edefcb3f5e1daa51067544a3b76"},
```
Comment (joshk) [inline mix.lock:62]: Updating `phoenix_test` is not possible right now. You will need to revert this update if you want the test suite to pass.

**Review by joshk — CHANGES_REQUESTED** [review]

File: mix.lock (hunk 1 above)
Comment (lawik) [reply mix.lock:62]: Ah, I had to nuke and pave deps for some ungodly state they got stuck in but happy to revert that.

**Comment by lawik** [comment]
@joshk this will be squashed and so on, but I think you should try the functionality, see how you like it :)

**Comment by lawik** [comment]
Should be good to go. Ideally take a look and see how you enjoy it @joshk :)

File: assets/js/app.js [hunk 2]
```diff
@@ -278,87 +298,79 @@ Hooks.WorldMap = {
...
+      fillOpacity: 1,
+      renderer: myRenderer
     };
 
-    // Clear previous defined device layer before adding markers
-    if (this.deviceLayer !== undefined) { this.map.removeLayer(this.deviceLayer); }
-
-    this.deviceLayer = L.geoJson(devices, {
-      pointToLayer: function (feature, latlng) {
+    var clusterLayer = L.markerClusterGroup({ chunkedLoading: true, chunkProgress: this.updateProgressBar });
```
Comment (nshoes) [inline assets/js/app.js:339]: This was difficult for me to read, so I ran through it and made some adjustments:

```js
var defaultOptions = {
  radius: 6,
  weight: 1,
  opacity: 0,
  fillOpacity: 1,
  renderer: myRenderer,
  fillColor: "#4dd54f"
}

var markerOfflineColor = "rgba(196,49,49,1)"
var markerOutdatedColor = "rgba(99,99,99,1)"

var devices = this.markers.reduce(function(acc, marker) {
  let location = marker["l"]
  var latLng = [location["at"], location["ng"]]

  // if no location or we don't care about mode, move on
  if (!location["ng"] && !location["at"]) return acc
  if (!["connected", "updated"].includes(mode)) return acc

  if (marker["s"] == "connected") {
    return L.circleMarker(latLng, defaultOptions)
  } else if (mode == "connected") {
    return L.circleMarker(latLng, Object.assign(defaultOptions, { fillColor: "rgba(196,49,49,1)" }))
  } else if (mode =="updated") {
    return L.circleMarker(latLng, Object.assign(defaultOptions, { fillColor: "rgba(99,99,99,1)" }))
  }
}, [])
```

It's not tested but it's close. Let me know what you think!

File: lib/mix/tasks/gen.devices.ex [hunk 3]
```diff
@@ -0,0 +1,41 @@
+defmodule Mix.Tasks.Gen.Devices do
```
Comment (nshoes) [inline lib/mix/tasks/gen.devices.ex:1]: I wrote something similar this morning seeding thousands of devices locally; this is nice!

File: lib/nerves_hub_web/live/dashboard/index.ex [hunk 4]
```diff
@@ -7,7 +7,7 @@ defmodule NervesHubWeb.Live.Dashboard.Index do
 
   alias Phoenix.Socket.Broadcast
 
-  @default_refresh 5000
+  @default_refresh 50000
```
Comment (nshoes) [inline lib/nerves_hub_web/live/dashboard/index.ex:10]: Any reason for picking 50 seconds? My brain wants this to be 45 or 60s 😆, but I don't feel strongly.

File: lib/nerves_hub_web/live/dashboard/index.ex [hunk 5]
```diff
@@ -117,8 +117,8 @@ defmodule NervesHubWeb.Live.Dashboard.Index do
 
   defp generate_map_marker(
          %{
-           id: id,
-           identifier: identifier,
+           id: _id,
+           identifier: _identifier,
```
Comment (nshoes) [inline lib/nerves_hub_web/live/dashboard/index.ex:121]:
```suggestion
```

File: assets/js/app.js [hunk 6]
```diff
@@ -237,15 +240,31 @@ Hooks.WorldMap = {
...
 
-    var mapOptionsNoZoom = {
+        // progressBar.classList.remove("hidden");
+        // progressBar.style.width = Math.round(processed / total * 100) + '%';
+      }
+
+      if (processed === total) {
+        // all markers processed - hide the progress bar:
+        // progress.classList.add = 'hidden';
+      }
```
Comment (nshoes) [inline assets/js/app.js:254]: Did this not work properly with your changes?

File: assets/js/app.js (hunk 2 above)
Comment (lawik) [reply assets/js/app.js:339]: good stuff, reworked a bit but a similar thing will be there :)

**Comment by lawik** [comment]
@nshoes looked at your notes, fixed all of it I believe :)

File: lib/nerves_hub_web/live/dashboard/index.ex (hunk 4 above)
Comment (nshoes) [reply lib/nerves_hub_web/live/dashboard/index.ex]: No no no please no I don't feel that strongly 😆😆😆

**Review by nshoes — APPROVED** [review]
🚀 🚀 🚀 🚀 🚀

**Comment by lawik** [comment]
Need to fix CI though

**Comment by nshoes** [comment]
@lawik done! :)

**Pull request merged**